### ISSUE TITLE
@singledispatch: pyright does not infer and check types correctly

### ISSUE THREAD

**Issue by mark-henry01** [microsoft/pyright#10375]
**Describe the bug**
In general `@singledispatch` allows to dispatch function/method invocations to different implementations based on provided argument types at runtime. But it also should allow pyright to infer argument and returns types conditionally based on static function signatures with function calls, which currently is not the case.

Expressed in other words, `@singledispatch` ideally should behave like `@overload` type-wise, while also auto-dispatching function calls at runtime (e.g. no need for multiple `isinstance` calls in function body). There are two limitations with pyright:
- no type checking (type and count of parameters) of `@singledispatch` functions/methods calls
- inferred function call return type is from default implementation, not from narrowest/closest function signature match

**Code or Screenshots**
```python
from functools import singledispatch
from typing import overload


@singledispatch
def dispatch(value: object) -> int | str:
    raise NotImplementedError(f"Unsupported type: {type(value)}")


@dispatch.register
def _(value: int) -> int:
    return value + 1


@dispatch.register
def _(value: str) -> str:
    return f"Hello {value}."


# works (pyright + runtime),
# but infers wide `int | str` return type
dispatch(10) 
dispatch("Hello")

# fails at runtime, but *no* errors with pyright
dispatch({"foo": "bar"})  # wrong argument type
dispatch(10, True)  # wrong argument count
```

In contrast, `@overload` produces stronger types (while lacking auto-dispatch):

```python
@overload
def ovload(value: int) -> int: ...


@overload
def ovload(value: str) -> str: ...


def ovload(value: int | str) -> str | int:
    if isinstance(value, int):
        return value + 1
    return f"Hello {value}."


# works (pyright + runtime)
# and infers narrow return type (good!)
ovload(3) # infers int return type
ovload("foo") # infers str return type

# pyright errors here (good!)
ovload({"foo": "bar"})
ovload(10, True)
```

**VS Code extension or command-line**
pyright 1.1.400, usage as language server in Neovim

Related:
https://github.com/microsoft/pyright/issues/7358
https://github.com/microsoft/pyright/issues/4472
https://github.com/microsoft/pyright/issues/988

**Comment by erictraut** [comment]
Pyright is working as intended here, so I don't consider this a bug.

The `functools. singledispatch` would require significant special-case logic within a type checker to model its behavior correctly. Pyright currently has no such logic. This is something that could theoretically be added, but there hasn't been sufficient demand up to this point to justify the (significant) work that would be required.

I'm going to change this to an enhancement request. I'm also going to close the issue for now because I don't have any current plans to support this functionality. I would consider reopening it in the future if there is sufficient signal (e.g. in the form of upvotes from other pyright users).

**Comment by Alexandre-SCHOEPP** [comment]
I would like this issue to be reopened @erictraut : I believe you could reuse a significant portion of the `overload` special case in here, especially if, for now, we limit ourselves to only the implementations defined in the same file as the dispatcher (which is the case in 90% of cases).

This is not a perfect solution by any stretch of the imagination, but it is immensely better than just falling off the type-hinting logic of Python entirely.
